de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,537 @@
+
+    /**
+     * Map for storing the set of resources shared for each shared reference
+     * point. Maps to <code>null</code> for completely shared reference points.
+     */
+    private Map<IReferencePoint, Set<IResource>> partiallySharedResourceMapping;
+
+    /**
+     * Set containing the currently completely shared reference points.
+     */
+    private Set<IReferencePoint> completelySharedReferencePoints;
+
+    /**
+     * Set containing the currently partially shared reference points.
+     */
+    private Set<IReferencePoint> partiallySharedReferencePoints;
+
+    public SharedReferencePointMapper() {
+        idToReferencePointMapping = new HashMap<String, IReferencePoint>();
+        referencePointToIDMapping = new HashMap<IReferencePoint, String>();
+        referencePointsOfUsers = new HashMap<User, List<String>>();
+        partiallySharedResourceMapping = new HashMap<IReferencePoint, Set<IResource>>();
+        completelySharedReferencePoints = new HashSet<IReferencePoint>();
+        partiallySharedReferencePoints = new HashSet<IReferencePoint>();
+    }
+
+    /**
+     * Adds a reference point to the set of currently shared reference points.
+     * <p>
+     * It is possible to "upgrade" a partially shared reference point to a
+     * completely shared reference point by just adding the same reference point
+     * with the same ID again that must now marked as not partially shared.
+     * </p>
+     * 
+     * @param id
+     *            the ID for the reference points
+     * @param referencePoint
+     *            the reference points to add
+     * @param isPartially
+     *            <code>true</code> if the reference point should be treated as
+     *            a partially shared reference point, <code>false</code> if it
+     *            should be treated as completely shared
+     * 
+     * @throws NullPointerException
+     *             if the ID or reference point is <code>null</code>
+     * @throws IllegalStateException
+     *             if the ID is already in use or the reference point was
+     *             already added
+     */
+    public synchronized void addReferencePoint(String id,
+        IReferencePoint referencePoint, boolean isPartially) {
+
+        CheckIfReferencePointIsAbleToBeShared(id, referencePoint);
+        if (isPartially) {
+            addReferencePointPartially(id, referencePoint);
+        } else if (partiallySharedReferencePoints.contains(referencePoint)) {
+            upgradeToCompletellyShared(id, referencePoint);
+            return;
+        } else {
+            addReferencePointCompletelly(id, referencePoint);
+        }
+
+        assert Collections.disjoint(completelySharedReferencePoints,
+            partiallySharedReferencePoints);
+
+        idToReferencePointMapping.put(id, referencePoint);
+        referencePointToIDMapping.put(referencePoint, id);
+
+        LOG.debug("added referencePoint " + referencePoint + " with ID " + id
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Define a constant instead of duplicating this literal " with ID " 3 times. ([squid:S1192](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1192))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=211&issues=AWVYURY9iZQY_aXE9GJL&open=AWVYURY9iZQY_aXE9GJL)

[ISSUE_KEY]: AWVYURY9iZQY_aXE9GJL